atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,294 @@
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<String> hostsWithEndpoints = ImmutableSet.of(newHostOne, oldHostOne);
+        setHostIds(
+                filterContainers(allHosts, server -> !hostsWithEndpoints.contains(server)), HostIdResult.softFailure());
+        setHostIds(filterContainers(allHosts, newHostOne::equalsIgnoreCase), HostIdResult.success(Set.of("uuid")));
+        setHostIds(filterContainers(allHosts, oldHostOne::equalsIgnoreCase), DEFAULT_RESULT);
+        assertThat(validator.getNewHostsWithInconsistentTopologies(
+                        filterServers(allHosts, newHosts::contains), allHosts))
+                .containsExactlyElementsOf(filterServers(allHosts, newHostOne::equalsIgnoreCase));
+    }
+
+    @Test
+    public void validateNewlyAddedHostsNoNewHostsAddedIfOldHostsDoNotHaveQuorum() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<CassandraServer> newCassandraServers = filterServers(allHosts, newHosts::contains);
+        Set<String> hostsOffline = ImmutableSet.of(oldHostOne, oldHostTwo);
+        setHostIds(filterContainers(allHosts, hostsOffline::contains), HostIdResult.hardFailure());
+        setHostIds(filterContainers(allHosts, server -> !hostsOffline.contains(server)), HostIdResult.success(uuids));
+        assertThat(validator.getNewHostsWithInconsistentTopologies(newCassandraServers, allHosts))
+                .containsExactlyElementsOf(newCassandraServers);
+    }
+
+    @Test
+    public void validateNewlyAddedHostsNoNewHostsAddedIfNewHostsDoNotHaveQuorum() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(newHosts);
+        Set<String> hostsOffline = ImmutableSet.of(newHostOne, newHostTwo);
+        setHostIds(filterContainers(allHosts, hostsOffline::contains), HostIdResult.hardFailure());
+        setHostIds(filterContainers(allHosts, server -> !hostsOffline.contains(server)), HostIdResult.success(uuids));
+        assertThat(validator.getNewHostsWithInconsistentTopologies(allHosts.keySet(), allHosts))
+                .containsExactlyElementsOf(allHosts.keySet());
+    }
+
+    @Test
+    public void validateNewlyAddedHostsAddsAllNewHostsIfCurrentServersDoNotHaveEndpoint() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        setHostIds(filterContainers(allHosts, newHosts::contains), HostIdResult.success(uuids));
+        setHostIds(filterContainers(allHosts, oldHosts::contains), HostIdResult.softFailure());
+        assertThat(validator.getNewHostsWithInconsistentTopologies(
+                        filterServers(allHosts, newHosts::contains), allHosts))
+                .isEmpty();
+    }
+
+    @Test
+    public void fetchHostIdsReturnsSoftFailureWhenException() throws Exception {
```

Comment:
Reminder to change this once you've cut an RC

Reply:
waiting on https://github.com/palantir/docker-proxy-rule/pull/543

Reply:
Boom! done.